docs/RustPOSIX/Architecture.md

```diff
@@ -47,6 +47,72 @@ The following are a breif description of the public methods
 - `insert_next_pipe` - given a pipe, insert it into the pipe table
 - `add_to_fd_table`, `rm_from_fd_table`, and `changedir` are **UNUSED!**
 
+# FILE SYSTEM
+# Inode Structure
+
+An important part of file representation, the Inode structure in SafePOSIX encompasses various formats of files supported by the system. Each file and directory in the filesystem has a different inode number.
+
+## Types of Inodes
+
+- **File**: Represents a standard file that contains data.
+- **CharDev**: Denotes a character device such as `/dev/null` or `/dev/zero`, including random number generators.
+- **Socket**: Represents a Unix domain socket, used for inter-process communication.
+- **Dir**: Represents a directory, organizing files and other directories.
+
+## Metadata within Inodes
+
+Each variant of Inode stores specific metadata:
+
+- **Common Metadata (all Inode types)**:
+  - `size`: Size of the file in bytes.
+  - `uid`, `gid`: User and group identity numbers.
+  - `mode`: Permissions of the file (read, write, execute).
+  - `linkcount`: Number of hard links that point to the inode.
+  - `refcount`: Counts open file descriptors of the file, allowing the system to track the file's lifetime internally.
+  - `atime`, `ctime`, `mtime`: Access, creation, modification timestamps.
+
+- **Directory-specific metadata (DirectoryInode)**:
+  - `filename_to_inode_dict`: A fast HashMap of filenames.
+
+- **Character Device-Specific Metadata (DeviceInode)**:
+  - `dev`: A DevNo struct holding the major and minor device numbers, uniquely identifying the device.
+
+## Ref Counting and File Lifetimes
+
+Refcount is used to manage file lifetimes. When a file is opened, its refcount increases; when closed, it decreases. The file is deleted only if both the link and ref count are 0, ensuring no current process can access the file.
+
+# Metadata
+
+## FilesystemMetadata Data Structure
+
+The FilesystemMetadata struct holds central filesystem-wide data:
+
+- `nextinode`: The next available inode number, ensuring uniqueness for new files and directories.
+- `dev_id`: The device ID where the filesystem resides.
+- `Inodetable`: HashMap mapping inode numbering to InodeEnum for quick lookups of file and directory metadata.
+
+# Logging for Correctness and Recovery
+
+## Write Ahead Logging
+
+- **Log File**: Modifications to filesystem metadata are initially recorded in the "lind.md.log" file before being committed to actual metadata structures.
```

Comment:
I think a little bit more here would also be useful. Espeically explaining *when* we log